libgalois/include/katana/GraphTopology.h

```diff
@@ -778,10 +826,29 @@ class KATANA_EXPORT EdgeTypeAwareTopology
   }
 
   bool has_transpose_state(
-      const EdgeShuffleTopology::TransposeKind& k) const noexcept {
+      const tsuba::RDGTopology::TransposeKind& k) const noexcept {
     return edge_shuff_topo_->has_transpose_state(k);
   }
 
+  auto adj_data() const noexcept { return per_type_adj_indices_.data(); }
+
+  auto dest_data() const noexcept { return edge_shuff_topo_->dest_data(); }
+
+  auto edge_property_index_data() const noexcept {
+    return topo().edge_property_index_data();
+  }
+
+  const EntityType* edge_type_index_data() const noexcept {
+    return edge_type_index_->index_to_type_map_data();
+  }
```

Comment:
This seems like a good time to bring up our encapsulation problem. I am certainly guilty of perpetuating it, but I think we should stop exposing pointers to class internals. I think one fundamental roadblock we are going to have is that these classes often don't have enough context to construct themselves. So we are going to need a way for outside code to manipulate the internals. I guess we can look for a more principled way to do it? If anyone has thoughts, I would love to noodle on this a little.

Reply:
agreed that this is gross. Unfortunately, I'm just following the existing pattern here. We don't have many options for getting data out of numarrays, and copying it needlessly would be quite painful.

Reply:
What about `PropertyGraphRetractor`? It seems to be our current container for leaky abstractions, maybe we could push the RDG data copying stuff into `PGR`?

Reply:
Bringing the RDGTopology creation into each topology helped a bit with this, but we can't completely remove these because of how the different GraphTopology Classes are defined.

`ShuffleTopology` inherits from `EdgeShuffleTopology` so here we could simply make the data pointers that `ShuffleTopology` needs from `EdgeShuffleTopology` protected and we would be okay.

But: `EdgeTypeAwareTopology` is different, it inherits from `BasicTopologyWrapper<EdgeShuffleTopology>` which means `EdgeTypeAwareTopology` can't get access to the data pointers of the `EdgeShuffleTopology` it needs to create an RDGTopology.